tests/api_test.py

```diff
@@ -3099,6 +3101,89 @@ def vjp(x_tangent):
+    x = np.ones([])
+    y = move(x)
+    self.assertDeleted(x)
+    self.assertEqual(y, 1.)
+
+  @jtu.skip_on_devices("cpu", "gpu")  # Buffer donation only supported on TPU.
+  def test_jit_donate_argnums_must_alias(self):
+    f = jit(lambda x: np.expand_dims(x, 0), donate_argnums=0)
+    with self.assertRaisesRegex(
+        ValueError,
+        r"donated but not used.*\n.*flat_args\[0\] \(spec: float32\[\]\)"):
+      f(np.ones([]))
+
+  @jtu.skip_on_devices("cpu", "gpu")  # Buffer donation only supported on TPU.
+  def test_jit_donate_argnums_static_argnums(self):
+    jit_fun = jit(lambda a, b, c, d: ((a + b + c), (a + b + d)),
+                  static_argnums=(0, 1), donate_argnums=(2, 3))
+
+    a = np.array(1)
+    b = np.array(2)
+    c = jax.device_put(np.array([1., 1.]))
+    d = jax.device_put(np.array([1., 1., 1.]))
+    e, f = jit_fun(a, b, c, d)
+    onp.testing.assert_allclose(e, np.array([4., 4.]))
+    onp.testing.assert_allclose(f, np.array([4., 4., 4.]))
+    self.assertNotDeleted(a)
+    self.assertNotDeleted(b)
+    self.assertDeleted(c)
+    self.assertDeleted(d)
+
+  def test_jit_nested_donate_ignored(self):
+    jit_fun = jit(lambda x: jit(lambda y: y ** 2, donate_argnums=0)(x))
+    a = jax.device_put(np.array(1))
+    with self.assertRaisesRegex(ValueError, "nested.*not supported"):
+      jit_fun(a)
+
+  def test_jvp_of_jit_donate_not_supported(self):
+    x = np.array(1.)
+    with self.assertRaisesRegex(ValueError,
+                                "JVP of.* jit.* which donates.* not supported"):
+      api.jvp(jit(lambda x: x ** 2, donate_argnums=0), (x,), (x,))
+
+  # === pmap ===
+
+  @jtu.skip_on_devices("cpu", "gpu")  # Buffer donation only supported on TPU.
+  def test_pmap_donate_argnums_invalidates_input(self):
+    # We can't just use `lambda x: x` because JAX simplifies this away.
+    move = pmap(lambda x: x + x - x, donate_argnums=0)
+    n = jax.local_device_count()
+    x = pmap(lambda x: x)(np.ones([n]))
+    y = move(x)
+    self.assertDeleted(x)
+    onp.testing.assert_allclose(y, [1.] * n)
+
+  @jtu.skip_on_devices("cpu", "gpu")  # Buffer donation only supported on TPU.
+  def test_pmap_donate_argnums_must_alias(self):
+    f = pmap(lambda x: np.expand_dims(x, 0), donate_argnums=0)
+    with self.assertRaisesRegex(
+        ValueError,
+        r"donated but not used.*\n.*flat_args\[0\] \(spec: float32\[\]\)"):
+      f(np.ones([1]))
+
+  def test_pmap_nested_donate_raises(self):
+    pmap_fun = jit(lambda x: pmap(lambda y: y ** 2, donate_argnums=0)(x))
+    a = pmap(lambda x: x)(np.array([1]))
+    with self.assertRaisesRegex(ValueError, "nested.*not supported"):
+      pmap_fun(a)
+
+  assertDeleted = lambda self, x: self._assertDeleted(x, True)
+  assertNotDeleted = lambda self, x: self._assertDeleted(x, False)
+
+  def _assertDeleted(self, x, deleted):
+    if hasattr(x, "device_buffer"):
+      self.assertEqual(x.device_buffer.is_deleted(), deleted)
```

Comment:
I am curious where is the logic that deletes the device buffer implemented?

Reply:
This happens in the XLA runtime when you pass a buffer to a computation that has input/output aliasing.

Reply:
Forgot to mention earlier, the upshot of marking deleted in Python is we can provide better error messages (instead of relying on errors from the runtime). But I think runtime errors are fine for now.

Reply:
I wouldn't block this PR on this, but I think it's important to give an error message couched in terms of JAX terminology (e.g. donation).

Anecdote: I tripped up into reusing a donated argument in a test a week after I patched this. It took me a while to realize that the errors I had were around reusing a donated buffer cuz the errors generated by the runtime are pretty oblique.
I signed up for this by being an early adopter, but ideally future users shouldn't also need to interpret TPU runtime error messages.

Reply:
When a buffer is donated and used as part of a computation it is marked as deleted and further uses will raise an error. We discussed leaving a tombstone on the buffer object to give better errors, I'll leave this for a follow up.